### PR TITLE
Add support for the ESP32-C6

### PR DESCRIPTION
**Pull request by jessebraham**
This is still a work in progress, however we're nearing completion so I think it's ready enough to start getting some eyes on it.

Unfortunately there were some hiccups with the git history, so things are a bit messy there; I've tried to fix things as much as possible but there are still some oddities.

At this point I have tested all examples, and all are functional except for the following:

- `clock_monitor`
- `i2s_sound`
- `twai`

The `twai` example is failing in the same way as the ESP32 and ESP32-S2, so we can likely hold off on completing this until #341 is resolved.

The `esp_hal_common::rtc_cntl` and `esp_hal_common::rtc_cntl::rtc` modules are still being worked on by @JurajSadel and will hopefully be completed soon; this will allow us to complete the `clock_monitor` example.

In addition to the above, I seem to be experiencing linker errors when trying to build any of the `embassy_*`, `pcnt_encoder`, or `ram` examples when using the `direct-boot` feature; they build fine without it. Not sure what's going on here.

Thank you to all who have contributed to this work, looking forward to getting it merged :)

### CODE REVIEW

**Comment by MabezDev** [comment]
FYI I tried looking at the direct boot errors but didn't actually run into any? Do you know how to reproduce? We should also make sure we apply the fix from #386 to the direct boot linker scripts too

**Comment by bjoernQ** [comment]
> FYI I tried looking at the direct boot errors but didn't actually run into any? Do you know how to reproduce? We should also make sure we apply the fix from #386 to the direct boot linker scripts too

I can at least reproduce linker errors for the `pcnt_encoder` and the `ram` example when using the direct-boot feature

**Comment by jessebraham** [comment]
@MabezDev running:

```bash
cargo build --release --example=embassy_hello_world --features=async,embassy,embassy-time-systick,direct-boot 
```

Results in:

```text
error: linking with `rust-lld` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/Users/jesse/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/bin:/Users/jesse/.pyenv/shims:/Users/jesse/.espressif/tools/xtensa-esp32s3-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-apple-darwin/bin:/Users/jesse/.espressif/tools/xtensa-esp32s2-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-apple-darwin/bin:/Users/jesse/.espressif/tools/xtensa-esp32-elf-gcc/8_4_0-esp-2021r2-patch3-x86_64-apple-darwin/bin:/Users/jesse/.espressif/tools/xtensa-lx106-elf/bin:/Users/jesse/.cargo/bin:/Users/jesse/.cargo:/Users/jesse/.pyenv/bin:/usr/local/opt/openjdk/bin:/Users/jesse/.local/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin:/Applications/Wireshark.app/Contents/MacOS" VSLANG="1033" "rust-lld" "-flavor" "gnu" "/var/folders/_9/z1ykrlk10t3c89lx0d80vx2c0000gn/T/rustc9G06cY/symbols.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.0.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.1.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.10.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.11.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.12.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.13.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.14.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.2.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.3.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.4.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.5.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.6.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.7.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.8.rcgu.o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f.embassy_hello_world.e744aa18-cgu.9.rcgu.o" "--as-needed" "-L" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps" "-L" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/release/deps" "-L" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/build/esp32c6-hal-31189cfd69fd103f/out" "-L" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/build/esp32c6-hal-31189cfd69fd103f/out" "-L" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/build/esp-riscv-rt-98e80c2aaec5eeaf/out" "-L" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/build/esp32c6-ac1c0b700a11b2c0/out" "-L" "/Users/jesse/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/riscv32imc-unknown-none-elf/lib" "-Bstatic" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp_println-f1516c68c9412db1.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp_backtrace-5a6289bef769b11f.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp_println-544ea99b727c1450.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp32c6_hal-3f896aea5bdaad97.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp_hal_common-6d0d2ec740686ed6.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libriscv_atomic_emulation_trap-d8b783ad643f9c42.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembassy_futures-108cf02e99f9bd00.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp32c6-7bb5fd2cd7cde49b.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libvcell-7ec1a82601af57ed.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembedded_can-b5ef1a17728ab62b.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libstrum-e0a2114bcfea621d.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembedded_hal_nb-4bb1c7cd27451fe8.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembedded_dma-7a1ad16d1fecbef4.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembedded_hal_async-7ab56fe3f173ceb6.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembassy_sync-0d51a6cc293c24bd.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libheapless-f3c0eb814efb41c9.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libstable_deref_trait-aba1dda1520b9a08.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libatomic_polyfill-3ecd6edd5c05a2bc.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libhash32-cb8cdea6ae15a045.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libbyteorder-3e593ad57c57f555.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembedded_hal-6e0af43c416e17a5.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libfugit-d3ee19f017842f16.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libgcd-21999270d138d20a.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libesp_riscv_rt-6ed04450734b273b.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libr0-c848c47157a64814.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libriscv-d3fddf532277f762.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libbit_field-72939deb9455fcce.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembassy_time-83f0d28cd6b88600.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembassy_executor-7e4a0c78bed41602.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libcfg_if-0223e26c41a1f9b3.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libstatic_cell-65bb6a6330a584a5.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembassy_time-03fb61eacce89eac.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libfutures_util-e0a4032f35e4f2fd.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libpin_project_lite-1fb41f722d72a765.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libfutures_task-82f39ec9889de3ae.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libpin_utils-dbfc4fe98ba9e9bc.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libfutures_core-b79b4bffe2585a3e.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libembedded_hal-985466426365a8fc.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libvoid-19380cf91179831d.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libnb-daeaec699b61b887.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libnb-9430177d6a7562a2.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libatomic_polyfill-bdf7ce7b5ed83cee.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libcritical_section-a298a3fe46a28bbe.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/librustc_std_workspace_core-027964a9f6098aa5.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libcore-e22379d5f534b18a.rlib" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/deps/libcompiler_builtins-5fc8f65b507cf7d5.rlib" "-Bdynamic" "-znoexecstack" "-L" "/Users/jesse/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/riscv32imc-unknown-none-elf/lib" "-o" "/Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/examples/embassy_hello_world-440a36bfd7244a5f" "--gc-sections" "-O1" "-Tlinkall.x"
  = note: rust-lld: error: /Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/build/esp32c6-hal-31189cfd69fd103f/out/riscv-link.x:108: unable to move location counter backward for: .rwtext.dummy
          rust-lld: error: /Users/jesse/Work/esp-hal/esp32c6-hal/target/riscv32imc-unknown-none-elf/release/build/esp32c6-hal-31189cfd69fd103f/out/riscv-link.x:108: unable to move location counter backward for: .rwtext.dummy
          rust-lld: warning: address (0x42000008) of section .text is not a multiple of alignment (256)
          rust-lld: error: section .rwtext.dummy at 0x40808814 of size 0xFFFFFFFFFFFFFFB8 exceeds available address space
          rust-lld: error: section .rwtext virtual address range overlaps with .bss
          >>> .rwtext range is [0x40808050, 0x40808813]
          >>> .bss range is [0x408087D0, 0x40808983]
          
          rust-lld: error: section .bss virtual address range overlaps with .rwtext.dummy
          >>> .bss range is [0x408087D0, 0x40808983]
          >>> .rwtext.dummy range is [0x40808814, 0x408087CB]
          
          rust-lld: error: section .rwtext load address range overlaps with .bss
          >>> .rwtext range is [0x8550, 0x8D13]
          >>> .bss range is [0x8CD0, 0x8E83]
          
          rust-lld: error: section .bss load address range overlaps with .rwtext.dummy
          >>> .bss range is [0x8CD0, 0x8E83]
          >>> .rwtext.dummy range is [0x8D14, 0x8CCB]
```

**Comment by jessebraham** [comment]
@MabezDev I think there's something funny going on with my system, let me try again.

**Comment by jessebraham** [comment]
Okay, this has been rebased now and includes all the latest `RTC` and linker script changes. Once a new version of `esp-backtrace` has been published we will need to use that, but all other dependencies are using published versions at this point.

I have tested a number of examples, all of which have worked, but have not done any comprehensive testing. This will likely require a group effort.

With all that said, I think this is pretty much as ready as it's going to get. There are a number of issues which still need resolving, however I will try to identify them so we can create issues at a later time; I'll update this comment when I am able.

File: esp32c6-hal/.cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+[target.riscv32imac-unknown-none-elf]
+runner = "espflash flash --monitor"
+rustflags = [
+  "-C", "link-arg=-Tlinkall.x",
+
+  # comment the cfgs below if you do _not_ wish to emulate atomics.
```
Comment (bjoernQ) [inline esp32c6-hal/.cargo/config.toml:6]: we don't need all of these overrides since `imac` already enables them

File: README.md [hunk 2]
```diff
@@ -4,21 +4,22 @@
...
-| [esp32c2-hal] |                    `riscv32imc-unknown-none-elf`                    |         [ESP32-C2]         |
-| [esp32c3-hal] |                    `riscv32imc-unknown-none-elf`                    |         [ESP32-C3]         |
-| [esp32s2-hal] |                      `xtensa-esp32s2-none-elf`                      |         [ESP32-S2]         |
-| [esp32s3-hal] |                      `xtensa-esp32s3-none-elf`                      |         [ESP32-S3]         |
+|     Crate     |            Target             | Technical Reference Manual |
+| :-----------: | :---------------------------: | :------------------------: |
+|  [esp32-hal]  |    `xtensa-esp32-none-elf`    |          [ESP32]           |
+| [esp32c2-hal] | `riscv32imc-unknown-none-elf` |         [ESP32-C2]         |
+| [esp32c3-hal] | `riscv32imc-unknown-none-elf` |         [ESP32-C3]         |
+| [esp32c6-hal] | `riscv32imc-unknown-none-elf` |         [ESP32-C6]         |
```
Comment (bjoernQ) [inline README.md:20]: riscv32imac-unknown-none-elf

**Comment by bjoernQ** [comment]
I tried all the examples we have for ESP32-C6 (only boot-loader mode). Not too bad but seems there are a few issues (some with the examples themselves and some with the driver, mostly clock related)

These are my results

adc.rs 🆗
advanced_serial.rs - baud rate is half (57600)
aes.rs 🆗
blinky.rs 🆗
blinky_erased_pins.rs - example code uses gpio 0 as a button but it's gpio 9
embassy_hello_world.rs 🆗
embassy_spi.rs - SCLK is 50kHz (i.e. it's half)
embassy_wait.rs - example code uses gpio 0 as a button but it's gpio 9
gpio_interrupt.rs 🆗
hello_rgb.rs 🆗
hello_world.rs  🆗
i2c_bmp180_calibration_data.rs 🆗
i2c_display.rs 🆗
i2s_read.rs - example descriptions mentions MCLK but it's not used, seems ok 🆗
i2s_sound.rs - tested with PCM5102, sounds okay in release mode 🆗
ledc.rs 🆗
mcpwm.rs 🆗
pcnt_encoder.rs - limits don't work?
pulse_control.rs 🆗
ram.rs 🆗
read_efuse.rs 🆗
rtc_watchdog.rs - seems it waits longer than 5 secs - so clocks seem to be off
serial_interrupts.rs - AT command detection not working
sha.rs 🆗
spi_eh1_device_loopback.rs - SCLK is 500kHz, should be 1000kHz
spi_eh1_loopback.rs - SCLK is 500kHz, should be 1000kHz
spi_loopback.rs - SCLK is 50kHz (i.e. it's half)
spi_loopback_dma.rs - SCLK is 50kHz (i.e. it's half)
systimer.rs 🆗
timer_interrupt.rs 🆗
usb_serial_jtag.rs - unable to get this to work - but might a problem with my OS since I can't even connect to the serial port
watchdog.rs 🆗

File: esp32c6-hal/.cargo/config.toml (hunk 1 above)
Comment (jessebraham) [reply esp32c6-hal/.cargo/config.toml]: Thanks, just realized I had copy-pasted this and not updated it last night 😅  Fixed now!

**Comment by jessebraham** [comment]
Building on top of @bjoernQ's testing, with my latest commits here is the current status of testing:

|           Example           | idf-bootloader | direct-boot |
| :-------------------------: | :------------: | :---------: |
|             adc             |       ✅       |     ✅      |
|       advanced_serial       |       ⏰       |     ⏰      |
|             aes             |       ✅       |     ✅      |
|           blinky            |       ✅       |     ✅      |
|     blinky_erased_pins      |       ❔       |     ❔      |
|     embassy_hello_world     |       ✅       |     ✅      |
|         embassy_spi         |       ⏰       |     ⏰      |
|        embassy_wait         |       ✅       |     ✅      |
|       gpio_interrupt        |       ✅       |     ✅      |
|          hello_rgb          |       ✅       |     ✅      |
|         hello_world         |       ✅       |     ✅      |
| i2c_bmp180_calibration_data |       ✅       |             |
|         i2c_display         |       ✅       |             |
|          i2s_read           |       ✅       |             |
|          i2s_sound          |       ✅       |             |
|            ledc             |       ❔       |     ❔      |
|            mcpwm            |       ✅       |     ✅      |
|        pcnt_encoder         |       ❔       |             |
|        pulse_control        |       ✅       |     ✅      |
|             ram             |       ✅       |     ✅      |
|         read_efuse          |       ✅       |     ✅      |
|        rtc_watchdog         |       ⏰       |     ⏰      |
|      serial_interrupts      |       ❌       |     ❌      |
|             sha             |       ✅       |     ✅      |
|   spi_eh1_device_loopback   |       ⏰       |     ⏰      |
|      spi_eh1_loopback       |       ⏰       |     ⏰      |
|        spi_loopback         |       ⏰       |     ⏰      |
|      spi_loopback_dma       |       ⏰       |     ⏰      |
|          systimer           |       ✅       |     ✅      |
|       timer_interrupt       |       ✅       |     ✅      |
|       usb_serial_jtag       |       ✅       |     ✅      |
|          watchdog           |       ✅       |     ✅      |

✅ = working as expected
⏰ = (suspected) clock issues
❔ = sort of working, but something is off
❌ = not working

I will update this table as I continue to test. Hopefully these remaining examples aren't too difficult to crack.

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+            1024,
+        );
+
+        // 100_000_000 is used to get rid of `float` calculations
+        let period = (100_000_000 * period_13q19 as u64) / (1 << RtcClock::CAL_FRACT);
+
+        (100_000_000 * 1000 / period) as u16
+    }
+
+    pub(crate) fn estimate_xtal_frequency() -> u32 {
```
Comment (MabezDev) [inline esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:673]: Should this function be called in the other stubbed out estimate_xtal_frequency function?

File: esp-hal-common/src/system.rs [hunk 4]
```diff
@@ -186,6 +194,69 @@ impl PeripheralClockControl {
...
+    pub fn enable(&mut self, peripheral: Peripheral) {
+        let system = unsafe { &*SystemPeripheral::PTR };
+
+        match peripheral {
+            Peripheral::Spi2 => {
+                system.spi2_conf.modify(|_, w| w.spi2_clk_en().set_bit());
+                system.spi2_conf.modify(|_, w| w.spi2_rst_en().clear_bit());
+            }
+            Peripheral::I2cExt0 => {
+                // FIXME
```
Comment (MabezDev) [inline esp-hal-common/src/system.rs:209]: Seems like this should be addressed before we merge.

File: esp-hal-common/src/system.rs [hunk 5]
```diff
@@ -186,6 +194,69 @@ impl PeripheralClockControl {
...
+            Peripheral::Ledc => {
+                system.ledc_conf.modify(|_, w| w.ledc_clk_en().set_bit());
+                system.ledc_conf.modify(|_, w| w.ledc_rst_en().clear_bit());
+            }
+            Peripheral::Mcpwm0 | Peripheral::Mcpwm1 => {
+                system.pwm_conf.modify(|_, w| w.pwm_clk_en().set_bit());
+                system.pwm_conf.modify(|_, w| w.pwm_rst_en().clear_bit());
+            }
+            Peripheral::ApbSarAdc => {
+                // TODO
```
Comment (MabezDev) [inline esp-hal-common/src/system.rs:226]: Ditto above.

File: esp-hal-common/src/uart.rs [hunk 6]
```diff
@@ -592,6 +592,65 @@ where
...
+        // we force the clock source to be APB and don't use the decimal part of the
+        // divider
+        let clk = clocks.apb_clock.to_Hz();
+        let max_div = 0b1111_1111_1111 - 1;
+        let clk_div = ((clk) + (max_div * baudrate) - 1) / (max_div * baudrate);
+
+        // UART clocks are configured via PCR
+        let pcr = unsafe { &*esp32c6::PCR::PTR };
+
+        match self.uart.uart_number() {
```
Comment (MabezDev) [inline esp-hal-common/src/uart.rs:606]: We have access the each Instances register block, so why do we need to match here?

**Review by MabezDev — CHANGES_REQUESTED** [review]
Little bit late, sorry! I found a few things I think we should address before merging :)

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs [hunk 7]
```diff
@@ -0,0 +1,697 @@
...
+            unsafe { ets_delay_us(50) };
+        } else {
+            pmu.hp_sleep_lp_ck_power
+                .modify(|_, w| w.hp_sleep_xpd_fosc_clk().clear_bit());
+        }
+    }
+
+    /// Get main XTAL frequency
+    /// This is the value stored in RTC register RTC_XTAL_FREQ_REG by the
+    /// bootloader, as passed to rtc_clk_init function.
```
[the remark this answers is not in the document]
Comment (bjoernQ) [reply esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:295]: seems like the 2nd stage bootloader just reads the value so it should be the ROM bootloader but probably we should just double check it

**Comment by bjoernQ** [comment]
> The blinky_erased_pins example works however it seems that only one in N button presses are actually detected; this is consistent between image formats. Probably not worth blocking on but there's something weird going on there.

That's exactly how the example works - it just reads the state of the button after the delay. So while it is not a good example of how to write real-world code it works as expected then.

**Comment by bjoernQ** [comment]
Seems I found the problem with `serial_interrupts` .... will prepare a PR for it

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bjoernQ) [reply esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:295]: Seems I was wrong - in direct-boot the read is invalid and we fall back to 40MHz which is anyway the only xtal frequency we currently support for C6

Since we need to know the xtal frequency anyway at build time (i.e. we have special features for 26MHz xtals for ESP32-C2 and ESP32 - for all others we only support 40MHz) we probably can just always use that known value

But probably nothing that needs to get addressed in this PR since it's like that for all targets and not really causing problems AFAICS

File: esp-hal-common/src/system.rs [hunk 8]
```diff
@@ -72,7 +80,7 @@ impl PeripheralClockControl {
             )
         };
 
-        #[cfg(any(esp32c2, esp32c3, esp32s2, esp32s3))]
+        #[cfg(any(esp32c2, esp32c3, esp32c6, esp32s2, esp32s3))]
```
Comment (bjoernQ) [inline esp-hal-common/src/system.rs:83]: here we already ruled out esp32c6 with the cfg on the impl block

**Comment by bjoernQ** [comment]
I fixed the DMA issues in direct-boot mode here: https://github.com/jessebraham/esp-hal/pull/7

File: esp-hal-common/src/timer.rs [hunk 9]
```diff
@@ -550,11 +566,27 @@ where
...
     pub fn new() -> Self {
+        #[cfg(esp32c6)]
+        Self::enable_clock();
         Self {
             phantom: PhantomData::default(),
         }
     }
 
+    #[cfg(esp32c6)]
+    fn enable_clock() {
```
[the remark this answers is not in the document]
Comment (JurajSadel) [reply esp-hal-common/src/timer.rs:577]: We are missing some peripherals for that PeripheralClockControl enable thing. For example, `sha`, `adc`, `usb-serial-jtag` don't take a mutable reference to PeripheralClockControl - we can address all of them in a separate PR.

File: README.md [hunk 10]
```diff
@@ -62,13 +65,15 @@ There are a number of other crates within the [esp-rs organization] which can be
 | [esp-backtrace] |                 Backtrace support for bare-metal applications                  |
 |  [esp-println]  |                Provides `print!` and `println!` implementations                |
 |  [esp-storage]  | Implementation of [embedded-storage] traits to access unencrypted flash memory |
+|   [esp-wifi]    |                Experimental `no_std` Wi-Fi/Bluetooth LE support                |
```
[the remark this answers is not in the document]
Comment (jessebraham) [reply README.md:68]: Just copy-pasting repo descriptions.

File: esp-hal-common/src/timer.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jessebraham) [reply esp-hal-common/src/timer.rs:577]: Why are we not adding the peripherals if they're missing? I feel like I'm missing something here.

File: esp-hal-common/src/uart.rs (hunk 6 above)
Comment (jessebraham) [reply esp-hal-common/src/uart.rs:608]: Because the register and field names are all different? I'm not sure I understand your question.

File: esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs (hunk 3 above)
Comment (jessebraham) [reply esp-hal-common/src/rtc_cntl/rtc/esp32c6.rs:673]: This should be removed altogether if we're removing the function in `rtc_cntl/mod.rs`, no?

File: esp-hal-common/src/system.rs (hunk 4 above)
Comment (jessebraham) [reply esp-hal-common/src/system.rs:209]: @bjoernQ would you mind looking at this? I can take a look but can't test anything. I think that the registers I expected weren't present or something, but they may be hiding elsewhere or be named differently.

**Comment by jessebraham** [comment]
I think at this point all examples are functioning as expected, both using the `idf-bootloader` format as well as `direct-boot`.

This PR is still far from perfect, but I think the remaining problems are isolated enough that we can create issues for them and address them individually; I'm not sure how much more we stand to gain by nitpicking this PR.

@bjoernQ and @MabezDev, let me know how you're feeling about the state of this PR, please.

@JurajSadel, @bjoernQ, and @MabezDev, thanks for all the work on this over the last couple months! It's been quite the learning experience for us and hopefully this will be easier in the future as a result.

File: esp-hal-common/src/uart.rs (hunk 6 above)
Comment (MabezDev) [reply esp-hal-common/src/uart.rs:608]: Ah yeah missed that, oops. I think we can abstract this better, but its fine for now I think.

**Comment by MabezDev** [comment]
I think this is in a good enough spot for merging! The few odd bits can be solved later :).
